File: finanze/application/use_cases/get_periodic_flows.py
from datetime import date, timedelta
from typing import Optional

from application.ports.periodic_flow_port import PeriodicFlowPort
from dateutil.relativedelta import relativedelta
from domain.earnings_expenses import FlowFrequency, PeriodicFlow
from domain.use_cases.get_periodic_flows import GetPeriodicFlows

frequency_delta_map = {
    FlowFrequency.MONTHLY: relativedelta(months=1),
    FlowFrequency.EVERY_TWO_MONTHS: relativedelta(months=2),
    FlowFrequency.QUARTERLY: relativedelta(months=3),
    FlowFrequency.EVERY_FOUR_MONTHS: relativedelta(months=4),
    FlowFrequency.SEMIANNUALLY: relativedelta(months=6),
    FlowFrequency.YEARLY: relativedelta(years=1),
}


def get_next_date(flow: PeriodicFlow) -> Optional[date]:
    if not flow.enabled:
        return None

    today = date.today()
    since_date = flow.since
    next_date = None

    if since_date > today:
        return since_date

    if flow.frequency == FlowFrequency.DAILY:
        next_date = today + timedelta(days=1)

    elif flow.frequency == FlowFrequency.WEEKLY:
        days_since = (today - since_date).days
        weeks_passed = days_since // 7
        next_date = since_date + timedelta(weeks=weeks_passed + 1)
        next_date = (
            next_date
            if next_date > today
            else since_date + timedelta(weeks=weeks_passed + 2)
        )

    elif flow.frequency in frequency_delta_map:
        delta = frequency_delta_map[flow.frequency]
        next_date = since_date
        while next_date <= today:
            next_date = next_date + delta

    if next_date and flow.until and next_date > flow.until:
        return None

    return next_date


class GetPeriodicFlowsImpl(GetPeriodicFlows):
    def __init__(self, periodic_flow_port: PeriodicFlowPort):
        self._periodic_flow_port = periodic_flow_port

    def execute(self) -> list[PeriodicFlow]:
        flows = self._periodic_flow_port.get_all()
        filled_flows = [
            PeriodicFlow(
                id=flow.id,
                name=flow.name,
                amount=flow.amount,
                currency=flow.currency,
                flow_type=flow.flow_type,
                frequency=flow.frequency,
                category=flow.category,
                enabled=flow.enabled,
                since=flow.since,
                until=flow.until,
                icon=flow.icon,
                linked=flow.linked,
                next_date=get_next_date(flow),
                max_amount=flow.max_amount,
            )
            for flow in flows
        ]
        return filled_flows

File: mobile/src/application/usecases/getPeriodicFlows.ts
import { FlowFrequency, PeriodicFlow } from "@/domain"

const frequencyMonthsMap: Partial<Record<FlowFrequency, number>> = {
  [FlowFrequency.MONTHLY]: 1,
  [FlowFrequency.EVERY_TWO_MONTHS]: 2,
  [FlowFrequency.QUARTERLY]: 3,
  [FlowFrequency.EVERY_FOUR_MONTHS]: 4,
  [FlowFrequency.SEMIANNUALLY]: 6,
  [FlowFrequency.YEARLY]: 12,
}

function parseLocalDate(dateStr: string): Date {
  // Expected: YYYY-MM-DD
  const [y, m, d] = dateStr.split("-").map(p => Number(p))
  return new Date(y, (m ?? 1) - 1, d ?? 1)
}

function formatLocalDate(date: Date): string {
  const y = date.getFullYear()
  const m = String(date.getMonth() + 1).padStart(2, "0")
  const d = String(date.getDate()).padStart(2, "0")
  return `${y}-${m}-${d}`
}

function startOfToday(): Date {
  const d = new Date()
  d.setHours(0, 0, 0, 0)
  return d
}

function daysInMonth(year: number, monthZeroBased: number): number {
  return new Date(year, monthZeroBased + 1, 0).getDate()
}

function addMonthsClamped(date: Date, months: number): Date {
  const year = date.getFullYear()
  const month = date.getMonth()
  const day = date.getDate()

  const total = month + months
  const newYear = year + Math.floor(total / 12)
  const newMonth = ((total % 12) + 12) % 12

  const dim = daysInMonth(newYear, newMonth)
  const newDay = Math.min(day, dim)

  return new Date(newYear, newMonth, newDay)
}

export function getNextDate(flow: PeriodicFlow): string | null {
  if (!flow.enabled) return null

  const today = startOfToday()
  const sinceDate = parseLocalDate(flow.since)

  if (sinceDate > today) {
    return formatLocalDate(sinceDate)
  }

  let nextDate: Date | null = null

  if (flow.frequency === FlowFrequency.DAILY) {
    nextDate = new Date(today)
    nextDate.setDate(nextDate.getDate() + 1)
  } else if (flow.frequency === FlowFrequency.WEEKLY) {
    const daysSince = Math.floor(
      (today.getTime() - sinceDate.getTime()) / (24 * 60 * 60 * 1000),
    )
    const weeksPassed = Math.floor(daysSince / 7)
    const candidate = new Date(sinceDate)
    candidate.setDate(candidate.getDate() + (weeksPassed + 1) * 7)

    if (candidate > today) {
      nextDate = candidate
    } else {
      const candidate2 = new Date(sinceDate)
      candidate2.setDate(candidate2.getDate() + (weeksPassed + 2) * 7)
      nextDate = candidate2
    }
  } else {
    const months = frequencyMonthsMap[flow.frequency]
    if (months) {
      let candidate = new Date(sinceDate)
      while (candidate <= today) {
        candidate = addMonthsClamped(candidate, months)
      }
      nextDate = candidate
    }
  }

  if (nextDate && flow.until) {
    const untilDate = parseLocalDate(flow.until)
    if (nextDate > untilDate) return null
  }

  return nextDate ? formatLocalDate(nextDate) : null
}
